hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -155,6 +156,9 @@ public BlockOutputStream(
     this.token = token;
 
     replicationIndex = pipeline.getReplicaIndex(pipeline.getClosestNode());
+    if (replicationIndex > 0) {
```

Comment:
I think it would be better to differentiate between Ratis and EC `BlockOutputStream` and `BlockOutputStreamEntryPool`. It can be more safe (Ratis code path shouldn't be changed) and more easy to test and better structured. I think this part can be added to a subclass (eg `ECBlockOutputStream`) which can override the required parts.

Reply:
I would tend to agree we should subclass BlockOutputStream here. The changes are quite small, but they may get larger in the future, and then there will be a temptation to add more and more EC things into the original class.

The HDFS code is full of "if EC else ..." logic and it makes it harder to follow the code generally. Its also good to avoid changing the original code if we can, as it works, and it makes reviews easier. Not just now, but later when we come to merge the branches. People get nervous when critical code paths get modified.